kratos/utilities/nurbs_utilities/snake_sbm_utilities.cpp

```diff
@@ -0,0 +1,872 @@
+                const auto& r_coords_true_boundary2 = i_cond.GetGeometry()[1].Coordinates();
+
+                std::vector<std::vector<double>> xy_coord_i_cond(2);
+                xy_coord_i_cond[0].resize(2); xy_coord_i_cond[1].resize(2); 
+                
+                xy_coord_i_cond[0][0] = r_coords_true_boundary1[0];
+                xy_coord_i_cond[1][0] = r_coords_true_boundary1[1];
+                xy_coord_i_cond[0][1] = r_coords_true_boundary2[0];
+                xy_coord_i_cond[1][1] = r_coords_true_boundary2[1];
+                
+                // Collect the intersections of the skin boundary with the knot values
+                std::vector<std::vector<int>> knot_span_uv(2);
+                knot_span_uv[0].resize(2); knot_span_uv[1].resize(2);
+
+                knot_span_uv[0][0] = (r_coords_true_boundary1[0]-starting_pos_uv[0]) / knot_step_uv[0]; // knot_span_u_1st_point
+                knot_span_uv[1][0] = (r_coords_true_boundary1[1]-starting_pos_uv[1]) / knot_step_uv[1]; // knot_span_v_1st_point
+                knot_span_uv[0][1] = (r_coords_true_boundary2[0]-starting_pos_uv[0]) / knot_step_uv[0]; // knot_span_u_2nd_point
+                knot_span_uv[1][1] = (r_coords_true_boundary2[1]-starting_pos_uv[1]) / knot_step_uv[1]; // knot_span_v_2nd_point
+
+                // In the inner case : check is the immersed object is inside the rectangular domain
+                if (is_inner && IsInside(knot_span_uv, n_knot_spans_uv))
+                    KRATOS_ERROR << "[SnakeSbmUtilities]:: The skin boundary provided is bigger than the background geometry in the parameter space." << std::endl;
+                
+                // In the outer case : additional check knot_span_uv computation on the domain border 
+                if (!is_inner)
+                {
+                    if (knot_span_uv[0][0] == n_knot_spans_uv[0]) knot_span_uv[0][0]--;
+                    if (knot_span_uv[1][0] == n_knot_spans_uv[1]) knot_span_uv[1][0]--;
+                    if (knot_span_uv[0][1] == n_knot_spans_uv[0]) knot_span_uv[0][1]--;
+                    if (knot_span_uv[1][1] == n_knot_spans_uv[1]) knot_span_uv[1][1]--;
+                }
+                
+                SnakeStep(r_skin_sub_model_part, knot_spans_available, id_matrix_knot_spans_available, 
+                         knot_span_uv, xy_coord_i_cond, knot_step_uv, starting_pos_uv);
+              
+                if (i_cond.GetGeometry()[1].Id() == id_first_node) {
+                    id_matrix_knot_spans_available++;
+                    new_inner_loop = true;
+                }
+            }
+        }
+
+        PointVector points;
+        for (auto &i_cond : r_skin_sub_model_part.Conditions()) {
+            points.push_back(PointTypePointer(new PointType(i_cond.Id(), i_cond.GetGeometry()[0].X(), i_cond.GetGeometry()[0].Y(), i_cond.GetGeometry()[0].Z())));
+        }
+        DynamicBins testBinInner(points.begin(), points.end());
```

Comment:
```suggestion
        DynamicBins testBinInner(points.begin(), points.end());
```
I think you should find a more descriptive name for `testBinInner`.